Move toolbar component to ui-classic and make it carbon

kavyanekkalapu opened the pull request:
Fixes: https://github.com/ManageIQ/manageiq-ui-classic/issues/7360

In this PR,moved toolbar component in react-ui-components to ui-classic repo, replaced PF components with carbon components, fixed accessibility and styling.

Before:

<img width="968" alt="Screen Shot 2020-12-01 at 3 42 14 PM" src="https://user-images.githubusercontent.com/37085529/100794672-e3f8da00-33eb-11eb-8c47-f6c7d94a63ae.png">

After:
<img width="1459" alt="Screen Shot 2020-12-01 at 2 06 08 PM" src="https://user-images.githubusercontent.com/37085529/100794690-eeb36f00-33eb-11eb-8f57-d0a400f2c560.png">
<img width="1380" alt="Screen Shot 2020-12-01 at 3 42 23 PM" src="https://user-images.githubusercontent.com/37085529/100794702-f1ae5f80-33eb-11eb-9fa5-2825eda853c9.png">
<img width="971" alt="Screen Shot 2020-12-01 at 2 06 26 PM" src="https://user-images.githubusercontent.com/37085529/100794714-f541e680-33eb-11eb-8243-6b89dde62eab.png">

@miq-bot add_reviewer @himdel ,@skateman

kavyanekkalapu commented:
testcases are failing due to changes in this pr, i am taking a look at that, other than that pr is ready for review.

himdel commented:
Do you have a screenshot of the two-state please?

The one I'm seeing on slack seems to have some vertical misalignment between twostate and the rest of the toolbar (the twostate icons were above the buttonSelect text line)

kavyanekkalapu commented:
@himdel you are right, looks like little mis alignment, will fix that

<img width="1538" alt="Screen Shot 2020-12-01 at 4 34 55 PM" src="https://user-images.githubusercontent.com/37085529/100799530-37225b00-33f3-11eb-97ca-eaca376b5b5f.png">

kavyanekkalapu commented:
@himdel pushed changes, latest screenshots

<img width="1647" alt="Screen Shot 2020-12-01 at 4 47 15 PM" src="https://user-images.githubusercontent.com/37085529/100800719-f75c7300-33f4-11eb-8182-7bcca9150fde.png">

<img width="1647" alt="Screen Shot 2020-12-01 at 4 47 15 PM" src="https://user-images.githubusercontent.com/37085529/100800736-fcb9bd80-33f4-11eb-8026-52745fc4ffc4.png">

himdel commented on app/javascript/menu/side-nav-menu-link.jsx:
```diff
@@ -2,6 +2,7 @@ import React, { forwardRef } from 'react';
 import PropTypes from 'prop-types';
 import { SideNavIcon, SideNavItem } from 'carbon-components-react/es/components/UIShell';
 import Link from 'carbon-components-react/es/components/UIShell/Link';
+import { OverflowMenu } from 'carbon-components-react/es/components/OverflowMenu';
```
`import OverflowMenu from...`

`node_modules/carbon-components-react/es/components/OverflowMenu/index.js` only exports a default export, not a named one

(should fix the tests :))

kavyanekkalapu commented:
Thanks @himdel testcases are fixed, pr is ready for review.

himdel commented:
There seems to be an issue where checking/unchecking items on the list screen no longer reliably updates the toolbar (based on onwhen).

I could get to a situation where only 1 vm i selected, but compare is still clickable (and fails on assert_rbac "/vm_infra/x_button?pressed=vm_compare", but I guess that's an artefact of the previous error).

himdel commented:
A minor visual issue with icon-only buttons (example from the VM compare screen):

![20201202161936](https://user-images.githubusercontent.com/289743/100900598-ea09bc00-34ba-11eb-862d-98a69f9e24bf.png)


the icon is not center aligned with the highlighted area.

himdel commented:
In Overview/Dashboard, there's still an old-looking button..

![Uploading 20201202162749.png…]()

himdel commented:
In Overview/Dashboard, there's still an old-looking button..

![20201202162749](https://user-images.githubusercontent.com/289743/100901037-584e7e80-34bb-11eb-9d0f-27be8a01ae36.png)

kavyanekkalapu commented:
> In Overview/Dashboard, there's still an old-looking button..
> 
> ![20201202162749](https://user-images.githubusercontent.com/289743/100901037-584e7e80-34bb-11eb-9d0f-27be8a01ae36.png)

i don't think dashboard page is using toolbar component , will fix dashboard page in another pr.

himdel commented:
We also seem to have lost custom buttons in generic objects, not yet sure if related to this PR yet, custom buttons do work just fine for VMs.

And there may be a `title` missing for disabled items.

himdel commented:
> And there may be a title missing for disabled items.

Yes, almost all the menu items (enabled or disabled) should have a title with a description.
For some disabled items, the reason why they are disabled is displayed in that title (feature not supported, no compatible provider, etc).

On master... (sorry, can't actually screenshot title popups under x11)...

![Untitled](https://user-images.githubusercontent.com/289743/100909092-2f7eb700-34c4-11eb-93d1-fe05835c869e.jpg)

kavyanekkalapu commented:
@himdel I tried to fix title for overflowmenu items,  i added code for it, but it has dependency of https://github.com/carbon-design-system/carbon/pull/7277, Carbon team mentioned to me that they are going to release version with this fix tomorrow, so once they released, title going to show up.
For disabled menu items, by default,  carbon styling has pointer events: none, so we can't hover, do you want me to remove that property?
Fixed icon only button padding as well
<img width="1053" alt="Screen Shot 2020-12-02 at 3 11 08 PM" src="https://user-images.githubusercontent.com/37085529/100926619-bd9d7200-34b1-11eb-9ef6-2f3d6ee2854b.png">

skateman commented on app/javascript/components/toolbar/tests/toolbar.test.js:
```diff
@@ -0,0 +1,97 @@
+import React from 'react';
```
Umm, I think this should be moved consistently to the place where any other test is for components...

skateman commented on app/javascript/components/toolbar/MenuIcon.js:
```diff
@@ -0,0 +1,21 @@
+    <span>{ props.text }</span>
+  </div>
+);
+MenuIcon.propTypes = {
```
Missing empty line

skateman commented on app/javascript/components/toolbar/Toolbar.jsx:
```diff
@@ -0,0 +1,123 @@
+      acc.splice(acc.length - 1, 0, i);
+    }
+    return acc;
+  }, [{ type: ButtonType.KEBAB, items: [] }]);
```
I guess it's been moved from the old code, but can't we do this in a nicer way, ideally immutable?

kavyanekkalapu replied:
@skateman yes it is just moved from old code, In this pr, i moved component from react repo to here as it is needed, changed PF components to CC and fix accessibility, unless there as a performance issue because of the way this code is written,  i don't want to make changes to existing working code. especially, for testing purposes, i am not really sure how and where this code is going to take an affect in ui, so want to keep the changes minimal and make it working.

skateman commented on app/javascript/components/toolbar/ToolbarList.jsx:
```diff
@@ -0,0 +1,85 @@
+      enabled,
+    };
+  });
+  return (
```
I guess we need a newline above this one...

skateman commented on app/javascript/components/toolbar/ToolbarList.jsx:
```diff
@@ -0,0 +1,85 @@
+  text: PropTypes.string,
+  title: PropTypes.string,
+};
+ToolbarList.defaultProps = {
```
Missing newline above...

skateman commented on app/javascript/components/toolbar/Toolbar.jsx:
```diff
@@ -0,0 +1,123 @@
+  } if (isButtonSelect(item) && (item.items.length > 0)) {
+    return <ToolbarList key={index} {...item} onClick={onClick} />;
+  }
+  return null;
```
Can we make these conditions nicer?

kavyanekkalapu replied:
@skateman yes it is just moved from old code, In this pr, i moved component from react repo to here as it is needed, changed PF components to CC and fix accessibility, unless there as a performance issue because of the way this code is written,  i don't want to make changes to existing working code. especially, for testing purposes, i am not really sure how and where this code is going to take an affect in ui, so want to keep the changes minimal and make it working.

kavyanekkalapu commented:
@skateman Fixed new line comments, moved testcases to common location where remaining test files lies.

skateman commented:
The main top-level button should be disabled if all its children are disabled:

![Screenshot from 2020-12-04 09-30-12](https://user-images.githubusercontent.com/649130/101141046-09682c80-3614-11eb-86ea-ba3f865852a6.png)


![Screenshot from 2020-12-04 09-34-45](https://user-images.githubusercontent.com/649130/101141043-0836ff80-3614-11eb-87e1-802053cd973a.png)

skateman commented:
Expand a dropdown menu that has enabled entities, hit `TAB` and the whole menu gets focused:
![Screenshot from 2020-12-04 09-42-29](https://user-images.githubusercontent.com/649130/101141900-3537e200-3615-11eb-900b-582d263be627.png)

After hitting `TAB` again the focus gets lost. I think the whole dropdown should not be focusable and the focus should go straight to the next right button in the toolbar instead as it would do if the dropdown is not open...

kavyanekkalapu commented:
> The main top-level button should be disabled if all its children are disabled:
> 
> ![Screenshot from 2020-12-04 09-30-12](https://user-images.githubusercontent.com/649130/101141046-09682c80-3614-11eb-86ea-ba3f865852a6.png)
> 
> ![Screenshot from 2020-12-04 09-34-45](https://user-images.githubusercontent.com/649130/101141043-0836ff80-3614-11eb-87e1-802053cd973a.png)
> 
> Also notice that the whole dropdown menu is being focused...

good catch, will fix this, thanks.

kavyanekkalapu commented:
> Expand a dropdown menu that has enabled entities, hit `TAB` and the whole menu gets focused:
> ![Screenshot from 2020-12-04 09-42-29](https://user-images.githubusercontent.com/649130/101141900-3537e200-3615-11eb-900b-582d263be627.png)
> 
> After hitting `TAB` again the focus gets lost. I think the whole dropdown should not be focusable and the focus should go straight to the next right button in the toolbar instead as it would do if the dropdown is not open...

@skateman i checked https://www.w3.org/TR/wai-aria-practices/#keyboard-interaction-13, according to this, we have to use up/down arrow , Escape  is closing menu and putting pack focus on selected menu button.  here  Nothing is mentioned about Tab, so this is working as expected. If AVT team raises TAB issue, we can always have a conversation showing this link.

kavyanekkalapu commented:
Fixed disable Toolabar Manu button.

Latest screenshot:
<img width="1151" alt="Screen Shot 2020-12-04 at 1 12 02 PM" src="https://user-images.githubusercontent.com/37085529/101199104-6d015280-3632-11eb-9c0c-5768163d6328.png">

skateman commented:
> @skateman i checked https://www.w3.org/TR/wai-aria-practices/#keyboard-interaction-13, according to this, we have to use up/down arrow , Escape is closing menu and putting pack focus on selected menu button. here Nothing is mentioned about Tab, so this is working as expected. If AVT team raises TAB issue, we can always have a conversation showing this link.

I'm okay with the up/down arrow usage, my problem is with the unwanted sideeffect. If the user hits tab after opening a menu, it all breaks...

kavyanekkalapu commented:
> > @skateman i checked https://www.w3.org/TR/wai-aria-practices/#keyboard-interaction-13, according to this, we have to use up/down arrow , Escape is closing menu and putting pack focus on selected menu button. here Nothing is mentioned about Tab, so this is working as expected. If AVT team raises TAB issue, we can always have a conversation showing this link.
> 
> I'm okay with the up/down arrow usage, my problem is with the unwanted sideeffect. If the user hits tab after opening a menu, it all breaks...

@skateman When we say our product is accessible, we are stating that it is accessible by  w3 standards https://www.w3.org/TR/wai-aria-practices/#keyboard-interaction-13, When user do something unusual than the component default keyboard interaction, we are not saying/claiming that it will still work, user should not be pressing tab inside of overflow menu, i checked couple of other accessible products that are using carbon react components(overflow), behavior is same there, anyway , i can write some hack and make it work but i feel it is unnecessary, Let me know your thoughts, if you still feel this is bug, i will fix it, thank you.

skateman commented:
Yes, IMO this is a bug...

himdel commented:
> For disabled menu items, by default, carbon styling has pointer events: none, so we can't hover, do you want me to remove that property?

Yes please, we've had the same problem with patternfly originaly, pointer-events:none prevents title from displaying.

kavyanekkalapu commented:
@himdel Fixed title on disabled item pointer events , updated carbon version which included title fix 
@skateman Fixed the tab issue, now when you further tab, focus is going to selected menu item, please review and let me know, thanks.

skateman commented:
Same thing happens, after hitting tab in an opened dropdown:
![Screenshot from 2020-12-08 16-19-07](https://user-images.githubusercontent.com/649130/101502396-2fbdfd00-3971-11eb-8d60-90a850b5075e.png)

h-kataria commented:
@skateman In my opinion this is working as expected, i am not sure if we should handle or change the behavior of tabbing as this is how it is working in other products and it's working as per ARIA specifications. User should not be tabbing thru overflow menus, they are only expected to use up/down arrow keys. To be able to get tabs working we will need to remove tabindex from the carbon overflowmenu by doing hack and that might cause other issues when tested with JAWS for A11Y. 
@gtanzillo what do you think?

gtanzillo commented:
I agree with Harpreet. Although it may not be ideal, it is a11y compliant and I think it would be best to go with this as is so that we can move on to other things.

skateman commented:
Well, but IF the user does it and TABs one more time, the focus is gone :thinking: :confused: that's my problem

kavyanekkalapu commented:
> Well, but IF the user does it and TABs one more time, the focus is gone 🤔 😕 that's my problem

@skateman have you checked latest code? i fixed that issue.

himdel commented:
@kavyanekkalapu It looks like all the title popups now show `[object Object]`. Sounds like you're passing an object instead of a string?

kavyanekkalapu commented:
> @kavyanekkalapu It looks like all the title popups now show `[object Object]`. Sounds like you're passing an object instead of a string?
> 
> (Correction: button groups are fine, but individual items show that object)

@himdel can you check now? i updated carbon-react version to 7.25.0, earlier i only updated base repo version, thanks.

himdel commented:
Yup, that did it, thanks :) all the toolbar titles are now :+1:

himdel commented:
Ok, looks like all issues are resolved :tada: :)
